src/engine/PDKind.cc

```diff
@@ -0,0 +1,295 @@
+    }
+    return VerificationResult(VerificationAnswer::UNKNOWN);
+}
+
+VerificationResult PDKind::solve(ChcDirectedGraph const & system) {
+    if (isTrivial(system)) {
+        return solveTrivial(system);
+    }
+    if (isTransitionSystem(system)) {
+        auto ts = toTransitionSystem(system);
+        return solveTransitionSystem(*ts);
+    }
+    SingleLoopTransformation transformation;
+    auto[ts, backtranslator] = transformation.transform(system);
+    assert(ts);
+    auto res = solveTransitionSystem(*ts);
+    return res;
+}
+
+/* Create a wrapper structure which will hold the result and also an induction invariant or a number of steps to a counter example */
+VerificationResult PDKind::solveTransitionSystem (TransitionSystem const & system) {
+    PTRef init = system.getInit();
+    PTRef query = system.getQuery();
+    PTRef transition = system.getTransition();
+
+    ReachabilityChecker reachability_checker(logic, system);
+
+    SMTConfig config;
+    TimeMachine tm{logic};
+    MainSolver solver(logic, config, "PDKIND");
+
+    MainSolver init_solver(logic, config, "Empty initial states");
+    init_solver.insertFormula(init);
+    { // Check for system with empty initial states
+        auto res = init_solver.check();
+        if (res == s_False) { return VerificationResult{VerificationAnswer::SAFE}; }
+
+        init_solver.insertFormula(query);
+        res = init_solver.check();
+        if (res == s_True) { return VerificationResult{VerificationAnswer::UNSAFE}; }
+    }
+
+    solver.insertFormula(init);
+
+    int n = 0;
+    PTRef p = logic.mkNot(query);
+    std::set<IFrameElement, IFrameElementCompare> inductionFrame;
+    inductionFrame.insert(IFrameElement(p, logic.mkNot(p)));
+
+    while (true) {
+        int k = n + 1; /* Pick k such that 1 <= k <= n + 1. (Create better approaches). */
+        auto res = push(system, inductionFrame, p, n, k, reachability_checker);
+
+        if (res.is_invalid) {
+            return VerificationResult(VerificationAnswer::UNSAFE, InvalidityWitness{});
+        }
+        if (res.i_frame == res.new_i_frame) {
+            // make k-inductive invariant as a conjunction of formulas in iframe and call kinductiontoinduction
+            return VerificationResult(VerificationAnswer::SAFE);
+        }
+        n = res.n;
+        inductionFrame = res.new_i_frame;
+    }
+}
+
+PushResult PDKind::push (TransitionSystem const & system, std::set<IFrameElement, IFrameElementCompare> & iframe, PTRef p, int n, int k, ReachabilityChecker & reachability_checker) {
+    std::size_t maxUnrollings = k;
+    PTRef init = system.getInit();
+    PTRef transition = system.getTransition();
+
+    SMTConfig config;
+    TimeMachine tm{logic};
+
+    std::queue<IFrameElement> q;
+    for (auto e : iframe) {
+        q.push(e);
+    }
+    std::set<IFrameElement, IFrameElementCompare> newIframe = {};
+    int np = n + k;
+    bool invalid = false;
+    while (!invalid && !q.empty()) {
```

Comment:
This is the preferred style
```suggestion
    while (not invalid and not q.empty()) {
```